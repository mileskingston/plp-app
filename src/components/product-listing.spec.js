import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import ProductListing from './product-listing';

describe('Product Listing Component', () => {
  let container = null;
  const repsonse = {
    breadCrumbTrailEntryView: [
      { label: "Electronics" },
      { label: "Tablets" }
    ],
    catalogEntryView: [{ 
        name: 'item',
        price: [{
          value: "300.0"
        }],
        thumbnail: '/image/url'
      },
      { 
        name: 'item 2',
        price: [{
          value: "20.0"
        }],
        thumbnail: '/image-2/url'
      }],
    facetView: [
      {
        value: "price_USD",
        name: "OfferPrice_USD",
        entry: [
          { label: "Name", value: "price_USD_Name", count: "2" },
          { label: "Name2", value: "price_USD_Name2", count: "2" },
        ],
      },
      {
        value: "mfName_ntk_cs",
        name: "ManufacturerName",
        entry: [
          { label: "Name", value: "mfName_ntk_cs_Name", count: "2" },
          { label: "Name2", value: "mfName_ntk_cs_Name2", count: "2" },
        ],
      }
    ],
    recordSetTotalMatches: 6,
  };

  beforeEach(() => {
    container = document.createElement("div");
    container.classList.add('test-wrapper');
    document.body.appendChild(container);
  });

  afterEach(() => {
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  it('renders component before data loaded', () => {
    render(<ProductListing />, container);
    expect(container).toMatchSnapshot();
  });

  it('renders component with data', async () => {
    jest.spyOn(global, 'fetch').mockImplementation(() =>
      Promise.resolve({
        json: () => Promise.resolve(repsonse)
      })
    );

    await act(async () => {
      render(<ProductListing />, container);
    });

    expect(container).toMatchSnapshot();
    global.fetch.mockRestore();
  });

  it('renders component when an error', async () => {
    const error = 'Somethings gone wrong';

    jest.spyOn(global, 'fetch').mockImplementation(() =>
      Promise.reject({ error: error })
    );

    await act(async () => {
      render(<ProductListing />, container);
    });

    expect(container).toMatchSnapshot();
    global.fetch.mockRestore();
  });
});