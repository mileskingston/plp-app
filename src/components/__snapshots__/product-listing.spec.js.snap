// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Listing Component renders component before data loaded 1`] = `
<div
  class="test-wrapper"
>
  <div
    class="product-listing"
  >
    <p>
      Loading...
    </p>
  </div>
</div>
`;

exports[`Product Listing Component renders component when an error 1`] = `
<div
  class="test-wrapper"
>
  <div
    class="product-listing"
  >
    
    <div
      class="filters"
    >
      <button
        class="button filters__button"
        type="button"
      >
        Filters
      </button>
      <div
        class="filters-facets filters-facets--hidden"
      />
    </div>
    <div
      class="product-grid"
    >
      <p>
        Sorry, no products
      </p>
    </div>
  </div>
</div>
`;

exports[`Product Listing Component renders component with data 1`] = `
<div
  class="test-wrapper"
>
  <div
    class="product-listing"
  >
    <h1
      class="title"
    >
      Tablets
       
      <span
        class="title__total"
      >
        6 Items
      </span>
    </h1>
    <div
      class="filters"
    >
      <button
        class="button filters__button"
        type="button"
      >
        Filters
      </button>
      <div
        class="filters-facets filters-facets--hidden"
      >
        <div
          class="facet"
        >
          <h3>
            Offer Price USD
          </h3>
          <label
            class="facet-entry"
            for="Name"
          >
            <input
              class="facet-entry__input"
              id="Name"
              name="Name"
              tabindex="0"
              type="checkbox"
              value="price_USD_Name"
            />
            <span
              class="facet-entry__label"
            >
              Name
               
              <span>
                2
              </span>
            </span>
          </label>
          <label
            class="facet-entry"
            for="Name2"
          >
            <input
              class="facet-entry__input"
              id="Name2"
              name="Name2"
              tabindex="0"
              type="checkbox"
              value="price_USD_Name2"
            />
            <span
              class="facet-entry__label"
            >
              Name2
               
              <span>
                2
              </span>
            </span>
          </label>
        </div>
        <div
          class="facet"
        >
          <h3>
            Manufacturer Name
          </h3>
          <label
            class="facet-entry"
            for="Name"
          >
            <input
              class="facet-entry__input"
              id="Name"
              name="Name"
              tabindex="0"
              type="checkbox"
              value="mfName_ntk_cs_Name"
            />
            <span
              class="facet-entry__label"
            >
              Name
               
              <span>
                2
              </span>
            </span>
          </label>
          <label
            class="facet-entry"
            for="Name2"
          >
            <input
              class="facet-entry__input"
              id="Name2"
              name="Name2"
              tabindex="0"
              type="checkbox"
              value="mfName_ntk_cs_Name2"
            />
            <span
              class="facet-entry__label"
            >
              Name2
               
              <span>
                2
              </span>
            </span>
          </label>
        </div>
      </div>
    </div>
    <div
      class="product-grid"
    >
      <div
        class="product"
      >
        <a
          href="/"
        >
          <img
            alt="item"
            class="product__image"
            src="https://www.demo.salmon-gcp.com/image/url"
          />
          <h2
            class="product__name"
          >
            item
          </h2>
          <p
            class="product__price"
          >
            $300.0
          </p>
        </a>
      </div>
      <div
        class="product"
      >
        <a
          href="/"
        >
          <img
            alt="item 2"
            class="product__image"
            src="https://www.demo.salmon-gcp.com/image-2/url"
          />
          <h2
            class="product__name"
          >
            item 2
          </h2>
          <p
            class="product__price"
          >
            $20.0
          </p>
        </a>
      </div>
    </div>
  </div>
</div>
`;
